### ISSUE TITLE
Import tensorflow pre-trained model using dnn modules encount an error

### ISSUE THREAD

**Issue by TerryBryant** [opencv/opencv#10123]
##### System information (version)
- OpenCV => 3.3.0
- Operating System / Platform => Windows 10(64 Bit)
- Compiler => Visual Studio 2015
- Tensorflow => 1.3.0
- Python => 3.6.1
##### Detailed description

I've wrote a lenet-5 model using tensorflow to train the mnist dataset, then I save the trained model file in .pb format. I use the dnn::readNetFromTensorflow() function to read the .pb file, but I get the following error:
```.txt
conv1/W_conv1/RandomStandardNormal:RandomStandardNormal(conv1/W_conv1/shape)
seed:[ ]
seed2:[ ]
dtype:0
T:0
OpenCV Error: Unspecified error (Unknown layer type RandomStandardNormal in op conv1/W_conv1/RandomStandardNormal) in cv::dnn::experimental_dnn_v1::`anonymous-namespace'::TFImporter::populateNet, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\tensorflow\tf_importer.cpp, line 883
```
Here are my tensorflow code of "conv1/W_conv1" part:
```.py
def weight_variable(shape, name):
    initial = tf.random_normal(shape, stddev=0.1, name=name)
    return tf.Variable(initial)
# conv1 layer
with tf.name_scope("conv1"):
    W_conv1 = weight_variable([5, 5, 1, 32], name='W_conv1') 
    b_conv1 = bias_variable([32], name='b_conv1')
    h_conv1 = tf.nn.relu(conv2d(x_image, W_conv1) + b_conv1, name='h_conv1') 
```

I couldn't find any problem in my tensorflow code, so I guess maybe my tensorflow version(released in Aug 17) is too updated for opencv 3.3.0(released in Aug 3). Hope anybody who can help me with this, thanks in advance.

**Comment by dkurt** [comment]
@TerryBryant, please tell us how you serialize the model. Is seems to me it isn't frozen by parameters.

**Comment by TerryBryant** [comment]
@dkurt Firstly I save the model in .ckpt format, then I transformed it to the .pb format, Here is the code:
```.py
sess = tf.Session()
saver.save(sess, "trained_model/mnist_cnn.ckpt")
tf.train.import_meta_graph('trained_model/mnist_cnn.ckpt.meta')
saver.restore(sess,tf.train.latest_checkpoint('trained_model/'))
tf.train.write_graph(sess.graph_def, save_dir, "Binary_Protobuf.pb", False)
```

**Comment by dkurt** [comment]
@TerryBryant, Binary_Peotobuf.pb has no weights. You need to merge it with checkpoint by freeze_graph tool. See https://www.tensorflow.org/extend/tool_developers/#freezing.

**Comment by gorbad-6** [comment]
Try to use the TF tools freeze_graph and transform_graph
```bash
bazel-bin/tensorflow/python/tools/freeze_graph --input_graph=save_dir/Binary_Protobuf.pb --input_checkpoint=trained_model/mnist_cnn --output_node_names=your_output_node_name --output_graph=your_output_graph
```
if that does not work,try to transform the output model with
```bash
bazel-bin/tensorflow/tools/graph_transforms/transform_graph
```
you can have a look at your model with
```bash
bazel-bin/tensorflow/tools/graph_transforms/summarize_graph
```

**Comment by TerryBryant** [comment]
@dkurt @gorbad-6 Thank you guys, I've fixed this problem

**Comment by TerryBryant** [comment]
Well, now I get another question as follows:
```.txt
OpenCV Error: Assertion failed (inpCn % ngroups == 0 && outCn % ngroups == 0) in cv::dnn::ConvolutionLayerImpl::getMemoryShapes, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\layers\convolution_layer.cpp, line 190
```
This is my opencv code:
```.cpp
Mat img = imread(imageFile, IMREAD_GRAYSCALE);
Size inputImgSize = cv::Size(28, 28);
if (inputImgSize != img.size())
    resize(img, img, inputImgSize);       //Resize image to input size
Mat inputBlob = blobFromImage(img);
net.setInput(inputBlob, "");
Mat result = net.forward();
```
By the way, I'm really confused with the inputs of net.setInput() and net.forward(), could you help me with this, @dkurt ?

**Comment by dkurt** [comment]
@TerryBryant, we need to know a graph definition to solve it.

**Comment by mrlittlepig** [comment]
I got the same error, well my graph is generated from tf.estimator.Estimator and I successfully freezed my graph by using freeze_graph.py, also I transformed my graph by
from tensorflow.tools.graph_transforms import TransformGraph
But, got the same error.
`IsVariableInitialized:IsVariableInitialized(global_step)
_output_shapes:[ ]
dtype:0
_class:[ ]
OpenCV Error: Unspecified error (Unknown layer type IsVariableInitialized in op IsVariableInitialized) in cv::dnn::experimental_dnn_v2::`anonymous-namespace'::TFImporter::populateNet, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\tensorflow\tf_importer.cpp, line 1296`
Please tell me how can I fix it.

**Comment by gorbad-6** [comment]
@mrlittlepig did you try to remove the Node with the op=IsVariableInitialized, with transform graph?
 eg run something like:
```bash
 bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=frozen_graph.pb --out_graph=frozen_transformed.pb --inputs=your_input_layer --outputs=your_output_layer/Activation --transforms="fold_batch_norms fold_constants sort_by_execution_order strip_unused_nodes remove_nodes(op=IsVariableInitialized)"
```

**Comment by gorbad-6** [comment]
@TerryBryant that error happens when the input shape of the Graph and the cv::Mat you pass are not aligned. Can you post the output of
```bash
bazel-bin/tensorflow/tools/graph_transforms/summarize_graph --in_graph=your_graph.pb
```

**Comment by mrlittlepig** [comment]
@gorbad-6 Thank you I will try it

**Comment by TerryBryant** [comment]
@gorbad-6 you can check out this link: 
/tools/graph_transforms/summarize_graph --in_graph=your_graph.pb
 @mrlittlepig
 
mrlittlepig commented 2 minutes ago
@gorbad-6 Thank you I will try it
@TerryBryant
  
            
 
Write

**Comment by TerryBryant** [comment]
@gorbad-6 well you can check out this link: https://github.com/TerryBryant/MyWork/tree/master/opencv_tensorflow
It contains all my code

**Comment by gorbad-6** [comment]
@TerryBryant I found two issues:
1) you did not export the graph itself into a pb. just add
```python
from tensorflow.python.framework import graph_io
graph_io.write_graph(sess.graph_def, "trained_model", "model.pb" )
```
to your code.
2) Another problem seems to be the dropout layer (mnist_cnn.py:65) since it has a tf.placeholder and TF's c++ can't deal with that yet. see eg. 
https://stackoverflow.com/questions/36804714/define-a-feed-dict-in-c-for-tensorflow-models/36817586#36817586
when i comment that layer out (i know dropouts are important, but you might solve it without placeholders) (mnist_cnn.py:65-68) and change the rest of the code accordingly, then when i freeze the graph with
```python
/tensorflow/bazel-bin/tensorflow/python/tools/freeze_graph --input_graph=model.pb --input_checkpoint=mnist_cnn.ckpt --output_node_names=softmax/prediction --output_graph=frozen_model.pb
```
then i can load the frozen model into opencv.

**Comment by TerryBryant** [comment]
@gorbad-6 According to your issues: 1, I do have transformed the ckpt files into .pb file using freeze_graph.py that you've told me, while I just don't mention it in the code. Code passes the readNetFromTensorflow function, but goes wrong in opencv_tensorflow.cpp:42. The logs are:
```.txt
OpenCV Error: Assertion failed (inpCn % ngroups == 0 && outCn % ngroups == 0) in cv::dnn::ConvolutionLayerImpl::getMemoryShapes, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\layers\convolution_layer.cpp, line 190
```
2, Dropout layer is indeed a problem and I'll try to fix it afterwards, so let's comment it temporarily. However, you still can't pass opencv_tensorflow.cpp:42, and receives the same error.

You can check my opencv_tensorflow.cpp, I think maybe there is something wrong with the variable "inputBlob".

**Comment by TerryBryant** [comment]
@gorbad-6 today I've updated all the code, now I can save the forzened .pb file even with out freeze_graph.py. Moreover, I've added a "python_tensorflow.py" to inference the "test8.png" and it gives the correct answer. However, I still can't pass the opencv_tensorflow.cpp:42 with the same error logs, exhausted...

**Comment by gorbad-6** [comment]
@TerryBryant the error above is, because in you pyhton code, you defined data as an output, i think. When I fix that, I get as the graph summary
```bash
/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/summarize_graph --in_graph=trained_model/frozen_model.pb 
Found 1 possible inputs: (name=x_input, type=float(1), shape=[?,784]) 
No variables spotted.
Found 1 possible outputs: (name=prediction, op=Softmax) 
Found 3274640 (3.27M) const parameters, 0 (0) variable parameters, and 0 control_edges
Op types used: 10 Const, 8 Identity, 4 Add, 3 Relu, 2 Conv2D, 2 MatMul, 2 MaxPool, 2 Reshape, 1 Placeholder, 1 Softmax
```
and then I can set the input blob as x_input. However, I am not sure OpenCv can reshape, and i get the assertion error you mentioned in the post before, no matter how i try to shape my image. Sadly, I do not have the resources to try to debug the opencv code, however you might. in that case, look at
opencv/modules/dnn/dnn.cpp:95 
insert a few prints and the the recompile opencv until you find out what goes wrong.

alternatively, look at #9992 and https://github.com/opencv/opencv/issues/10065#issuecomment-343427950

**Comment by TerryBryant** [comment]
@gorbad-6 hi, are you still there? Now I've changed my input data format and managed to use the opencv to import tensorflow model. However, when I compare the inference result between opencv and tensorflow, they are totally different(check https://github.com/TerryBryant/MyWork/blob/master/opencv_tensorflow/python_tensorflow.py), have you faced similar conditions like that?

**Comment by gorbad-6** [comment]
@TerryBryant sorry, I have no idea what the problem could be. The only thing I do not quite understand in code is why you set 'input/input_x' as output node names (mnist_cnn.py: 86)

**Comment by TerryBryant** [comment]
@gorbad-6 'input/input_x' is not an output(python_tensorflow.py: 47 and 51), it's an input tensor. Well, guess I have to put this question on the stackoverflow. Anyway, thank you for your patience in participating in the discussion.

**Comment by Dxye** [comment]
@TerryBryant  Have you successfully make prediction using opencv dnn? I met an error at net.forward. 

OpenCV Error: Assertion failed (mapIt != reuseMap.end()) in cv::dnn::experimental_dnn_v1::BlobManager::numReferences, file D:\Workspace\libraries\opencv-3.3.0\modules\dnn\src\dnn.cpp, line 419

I searched but seems no one met this error.
